c++/cpp2py/converters/vector.hpp

```diff
@@ -5,63 +5,60 @@
 
 #include "../macros.hpp"
 #include "../numpy_proxy.hpp"
+#include "../py_converter.hpp"
 
 namespace cpp2py {
 
   template <typename T> static void delete_pycapsule(PyObject *capsule) {
-    auto *ptr = static_cast<std::unique_ptr<T[]> *>(PyCapsule_GetPointer(capsule, "guard"));
+    auto *ptr = static_cast<T *>(PyCapsule_GetPointer(capsule, "guard"));
     delete ptr;
   }
 
   // Convert vector to numpy_proxy, WARNING: Deep Copy
-  template <typename T> numpy_proxy make_numpy_proxy_from_vector(std::vector<T> const &v) {
-
-    auto *data_ptr = new std::unique_ptr<T[]>{new T[v.size()]};
-    std::copy(begin(v), end(v), data_ptr->get());
-    auto capsule = PyCapsule_New(data_ptr, "guard", &delete_pycapsule<T>);
-
-    return {1, // rank
-            npy_type<std::remove_const_t<T>>,
-            (void *)data_ptr->get(),
-            std::is_const_v<T>,
-            v_t{static_cast<long>(v.size())}, // extents
-            v_t{sizeof(T)},                   // strides
-            capsule};
+  template <typename T> numpy_proxy make_numpy_proxy_from_vector(std::vector<T> v) {
+
+    if constexpr (has_npy_type<T>) {
+      auto *vec_heap = new std::vector<T>{std::move(v)};
+      auto capsule   = PyCapsule_New(vec_heap, "guard", &delete_pycapsule<std::vector<T>>);
+
+      return {1, // rank
+              npy_type<std::remove_const_t<T>>,
+              (void *)vec_heap->data(),
+              std::is_const_v<T>,
+              v_t{static_cast<long>(vec_heap->size())}, // extents
+              v_t{sizeof(T)},                           // strides
+              capsule};
+    } else {
+      std::vector<pyref> vobj(v.size());
+      std::transform(begin(v), end(v), begin(vobj), [](auto &&x) { return py_converter<std::decay_t<T>>::c2py(std::move(x)); });
+      return make_numpy_proxy_from_vector(std::move(vobj));
+    }
   }
 
   // Make a new vector from numpy view
   template <typename T> std::vector<T> make_vector_from_numpy_proxy(numpy_proxy const &p) {
     EXPECTS(p.extents.size() == 1);
-    EXPECTS(p.strides == v_t{sizeof(T)});
 
-    T *data   = static_cast<T *>(p.data);
     long size = p.extents[0];
-
     std::vector<T> v(size);
-    std::copy(data, data + size, begin(v));
+
+    if (p.element_type == npy_type<pyref>) {
+      auto **data = static_cast<PyObject **>(p.data);
```

Comment:
same as before

Reply:
Which previous comment are you referring to?